### DISCUSSION TITLE
Instance of 를 지양해야 하는 이유에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by nayonsoso** [2024-woowacourse-study/level-interview#69] [OOP]
.

**Comment by alstn113** [comment]
"instanceof" 연산자를 지양해야 하는 이유는 다음과 같다.

1. instanceof 연산자를 사용하면 코드가 특정 클래스나 타입에 의존하게 된다. 이는 유연성이 떨어지며, 새로운 클래스나 타입이 추가될 때 기존 코드를 변경해야 할 수 있습니다. 
2. 객체 지향 프로그래밍에서는 다형성을 통해 동일한 인터페이스를 구현하는 다양한 객체를 동일한 방식으로 처리할 수 있습니다. instanceof 연산자는 이러한 다형성을 무시하고 각 객체의 타입을 직접 확인하므로 객체 지향 프로그래밍의 원칙을 위반한다. 
3. 코드의 가독성이 떨어지고, 코드의 의도가 명확하지 않을 수 있다. 또한 다양한 객체 유형에 대해 동일한 코드를 반복 작성해야 하므로 코드의 유지보수성이 저하될 수 있습니다. 
4. 객체가 특정 클래스나 타입에 의존하게 되어 객체 간의 유연한 상호작용이 어려워질 수 있습니다.

이러한 이유로 가능하면, instanceof 연산자를 지양하고, 대신에 다형성과 상속을 활용하여 객체 간의 관계를 명확하게 정의하는 것이 좋습니다.

**Comment by GIVEN53** [comment]
1. 유연성과 확장성의 저하
    - 특정 클래스나 하위 클래스에 의존하게 되므로 유연성이 저하되며, 새로운 하위 클래스를 추가할 때 코드를 수정해야 한다.
2. 코드의 복잡성 증가
    - 다른 클래스의 유형을 체크하고 그에 따른 로직을 추가해야 한다.
3. 객체 지향 원칙 위배
    - 다형성을 무시하고 객체의 구체적인 타입에 의존하는 코드를 작성해야 한다.
4. 객체 설계의 재고
    - 결국 설계의 문제일 수 있다. 객체의 역할과 책임을 정확하게 구분하고, 각 객체가 스스로 책임을 수행할 수 있도록 설계해야 한다.
